cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeMonitor.java

```diff
@@ -225,74 +296,124 @@ protected void run() {
 
-    // Publish the remaining program state messages. We should retry indefinitely
-    String topicConfig = Constants.AppFabric.PROGRAM_STATUS_EVENT_TOPIC;
-    String topic = requestKeyToLocalTopic.get(topicConfig);
-    try {
-      Retries.callWithRetries(() -> publish(topicConfig, topic, lastProgramStateMessages),
-                              RetryStrategies.fixDelay(1, TimeUnit.SECONDS), Retries.ALWAYS_TRUE);
-    } catch (Exception e) {
-      // This shouldn't happen since we have infinite retry.
-      LOG.error("Failed to publish remaining program state messages for program run {}", programRunId, e);
+    if (!lastProgramStateMessages.isEmpty()) {
+      String topicConfig = Constants.AppFabric.PROGRAM_STATUS_EVENT_TOPIC;
+      String topic = requestKeyToLocalTopic.get(topicConfig);
+
+      // Publish last program state messages with Retries
+      Retries.runWithRetries(() -> Transactionals.execute(transactional, context -> {
+        AppMetadataStore store = AppMetadataStore.create(cConf, context, dsFramework);
+        publish(topicConfig, topic, lastProgramStateMessages, store);
+
+        // cleanup all the offsets after publishing terminal states.
+        for (String topicConf : requestKeyToLocalTopic.keySet()) {
+          store.deleteSubscriberState(topicConf, programRunId.getRun());
+        }
+      }), retryStrategy);
     }
   }
 
+  private Map<String, MonitorConsumeRequest> initTopics() {
+    Map<String, MonitorConsumeRequest> consumeRequests = new HashMap<>();
+
+    Transactionals.execute(transactional, context -> {
+      AppMetadataStore store = AppMetadataStore.create(cConf, context, dsFramework);
+
+      for (String topicConfig : requestKeyToLocalTopic.keySet()) {
+        String messageId = store.retrieveSubscriberState(topicConfig, programRunId.getRun());
+        consumeRequests.put(topicConfig, new MonitorConsumeRequest(messageId, limit));
+      }
+    });
+
+    return consumeRequests;
+  }
+
   /**
    * Processes the given set of monitoring messages.
    *
    * @param monitorResponses set of messages to be processed.
    * @return the latest message publish time or {@code -1L} if there is no message to process
    */
-  private long processResponse(Map<String, List<MonitorMessage>> monitorResponses) throws Exception {
+  private long processResponse(Map<String, List<MonitorMessage>> monitorResponses,
+                               AppMetadataStore store) throws Exception {
     long latestPublishTime = -1L;
 
     for (Map.Entry<String, List<MonitorMessage>> monitorResponse : monitorResponses.entrySet()) {
       String topicConfig = monitorResponse.getKey();
       List<MonitorMessage> monitorMessages = monitorResponse.getValue();
 
-      if (monitorMessages.isEmpty()) {
+      // Skip terminal program status messages to delay the deprovisioning.
+      if (monitorMessages.isEmpty() ||
+              (topicConfig.equals(Constants.AppFabric.PROGRAM_STATUS_EVENT_TOPIC) && programFinishTime > 0)) {
```

Comment:
Why need to change the logic here? This change means we'll be fetching the messages again instead of buffering them?

Reply:
We are updating the `lastProgramStateMessages` in `updateProgramFinishTime`. So we skip them here so that they are not published.

Reply:
I have refactored `processResponse` to update `lastProgramStateMessages` when terminal program state is detected.

Reply:
Generally, it's best to minimize number of methods that will mutate global states (aka fields). Better to only have the update happening in one place, while other methods takes states and return new states.

Reply:
Now There are only 2 methods updating states: 
 `updateProgramFinishTime` : Responsible for updating ProgramFinishTime and in memory `lastProgramStateMessages`
`updateConsumeRequests`: Responsible for updating in memory offsets for other topics.

I have added reason for having 2 update methods in the code. 

Please let me know your thoughts on this implementation.